### PR TITLE
Fix variable usage inside and/2 during in/2 expansion

### LINKED ISSUE #5256: Unclear compilation error "cannot define variable _ inside and".
### Environment

* Elixir version (elixir -v):

Erlang/OTP 19 [erts-8.1] [source] [64-bit] [smp:4:4] [async-threads:10] [kernel-poll:false]

IEx 1.4.0-dev (662ecf9)

* Operating system:
Ubuntu 16.04

### Current behavior

When compiling this code snippet:
```
        if response.result == :failure and
           response.content.code in 300..399 and
           cmd.state.recurse do
```

I get the following error (which also does not show a line number, so it took me quite some time to find the offending statement):
```
== Compilation error on file lib/dialplan/command/proxy.ex ==
** (CompileError) lib/dialplan/command/proxy.ex: cannot define variable _ inside and
    (stdlib) lists.erl:1338: :lists.foreach/2
    (stdlib) erl_eval.erl:670: :erl_eval.do_apply/6
```

On Elixir 1.3.3 it compiles fine, and if I change the code to either
```
        if response.result == :failure and
           response.content.code >= 300 and
           response.content.code <= 400 and
           cmd.state.recurse do
```
or
```
        code = response.content.code
        if response.result == :failure and
           code in 300..399 and
           cmd.state.recurse do
```
It compiles without errors.

### Expected behavior

I expect this code to be compiled, just like on 1.3.3.

### PR DESCRIPTION
**Pull request by lexmag**
Closes #5256.

### CODE REVIEW

File: lib/elixir/test/elixir/kernel_test.exs [hunk 1]
```diff
@@ -233,20 +233,30 @@ defmodule KernelTest do
...
+    expected =
+      "(fn var -> " <>
+      ":erlang.or(:erlang.=:=(var, 1), :erlang.=:=(var, 2)) " <>
+      "end).(rand())"
+    assert expand_to_string(quote(do: rand() in [1 | [2]])) == expected
+  end
+
+  response = %{code: 200}
+  if is_map(response) and response.code in 200..299 do
+    :pass
```
Comment (whatyouhide) [inline lib/elixir/test/elixir/kernel_test.exs]: 😕 ? ✂️ ? :)

**Comment by josevalim** [comment]
After reviewing this pull request, one other option is to compile "and" to
a case when outside of guards. This would also fix issues with any other
code that may do the same. Thoughts?

-- 


*José Valim*
www.plataformatec.com.br
Skype: jv.ptec
Founder and Director of R&D

File: lib/elixir/test/elixir/kernel_test.exs (hunk 1 above)
Comment (lexmag) [inline lib/elixir/test/elixir/kernel_test.exs]: It's actually the test for the reported issue. :bowtie:

Comment (whatyouhide) [inline lib/elixir/test/elixir/kernel_test.exs]: @lexmag yep, is it outside of a `test` for a reason though? :|

Comment (lexmag) [inline lib/elixir/test/elixir/kernel_test.exs]: @whatyouhide it needs to be outside of the function context, we can move it into a test but still with a module definition wrapping.

Comment (whatyouhide) [inline lib/elixir/test/elixir/kernel_test.exs]: It just seems weird outside of a `test` (without any comments especially 😛) and I assume failures in such test would not be ideal, so personally I'd go with the module def approach 👍

**Comment by lexmag** [comment]
Closing in favour of #5295.

**Pull request closed without merging**